docs/style-guide.rst

```diff
@@ -51,126 +50,952 @@ into the making of Hy.
+    (foo (bar 2))
+
+    ;; BAD
+    [ 1 2 3 ]
+    ( foo ( bar 2 ) )
+
+    ;; BAD. And ugly.
+    [ 1 2 3]
+    (foo( bar 2) )
+
+Grouping
+--------
+
+Use whitespace to show implicit groups, but be consistent within a form.
+
+.. code-block:: clj
+
+    ;; Older Lisps would typically wrap such groups in even more parentheses.
+    ;; (The Common Lisp LOOP macro was a notable exception.)
+    ;; But Hy takes after Clojure, which has a lighter touch.
+
+    ;; BAD. Can't tell key from value without counting
+    {1 9 2 8 3 7 4 6 5 5}
+
+    ;; PREFERRED. This can fit on one line. Clojure would have used commas
+    ;; here, but those aren't whitespace in Hy. Use extra spaces instead.
+    {1 9  2 8  3 7  4 6  5 5}
+
+    ;; OK. And preferred if it couldn't fit on one line.
+    {1 9
+     2 8
+     3 7
+     4 6
+     5 5}  ; Newlines show key-value pairs in dict.
+
+    ;; BAD
+    ;; This grouping makes no sense.
+    #{1 2
+      3 4}  ; It's a set, so why are there pairs?
+
+    ;; BAD
+    ;; This grouping also makes no sense. But, it could be OK in a macro or
+    ;; something if this grouping was somehow meaningful there.
+    [1
+     1 2
+     1 2 3]  ; wHy do you like random patterns? [sic pun, sorry]
+
+    ;; Be consistent. Separate all groups the same way in a form.
+
+    ;; BAD
+    {1 9  2 8
+     3 7  4 6  5 5}  ; Pick one or the other!
+
+    ;; BAD
+    {1 9  2 8 3 7  4 6  5 5}  ; You forgot something.
+
+    ;; Groups of one must also be consistent.
+
+    ;; PREFERRED
+    (foo 1 2 3)  ; No need for extra spaces here.
+
+    ;; OK, but you could have fit this on one line.
+    (foo 1
+         2
+         3)
+
+    ;; OK, but you still could have fit this on one line.
+    [1
+     2]
+
+    ;; BAD
+    (foo 1 2  ; This isn't a pair?
+         3)  ; Lines or spaces--pick one or the other!
+
+    ;; PREFERRRED
+    (foofunction (make-arg)
+                 (get-arg)
+                 #tag(do-stuff)  ; Tags belong with what they tag.
+                 #* args  ; #* goes with what it unpacks.
+                 :foo spam
+                 :bar eggs  ; Keyword args are also pairs. Group them.
+                 #** kwargs)
+
+    ;; PREFERRED. Spaces divide groups on one line.
+    (quux :foo spam  :bar eggs  #* with-spam)
+    {:foo spam  :bar eggs}
+
+    ;; OK. The colon is still enough to indicate groups.
+    (quux :foo spam :bar eggs #* with-spam)
+    {:foo spam :bar eggs}
+    ;; OK.
+    ("foo" spam "bar" eggs}
+
+    ;; BAD. Can't tell key from value.
+    (quux :foo :spam :bar :eggs :baz :bacon)
+    {:foo :spam :bar :eggs :baz :bacon}
+    {"foo" "spam" "bar" "eggs" "baz" "bacon"}
+
+    ;; PREFERRED
+    (quux :foo :spam  :bar :eggs  :baz :bacon)
+    {:foo :spam  :bar :eggs  :baz :bacon}
+    {"foo" "spam"  "bar" "eggs"  "baz" "bacon"}
+
+    ;; OK. Yep, those are pairs too.
+    (setv x 1
+          y 2)
+
+    ;; PREFERRED. This fits on one line.
+    (setv x 1  y 2)
+
+    ;; BAD. Doesn't separate groups.
```

Comment:
nit: there's no preferred example for the if, only "ok" so it's not clear if there's actually a preference for something else. I'm guessing preferred would be (based on the previous examples):

```hy
(print (if (< n 0.0) "negative"
           (= n 0.0) "zero"
           ...))
```

Reply:
It's been a while since I wrote this, so I don't remember exactly why there's no example for that.

Unfortunately, other Lisps don't completely agree on how to indent `if`s (The three laws always apply, of course.), so I'm not sure what the preference should be. I suppose I can pick one and be more explicit about it, as I said last year, 
> I would still consider most of this tentative even if merged, but the community might not have even thought of some of these issues, so I think we can have a more productive discussion with a framework to start from.

Emacs style dedents the else clause slightly.
```
;; OK?
(if (predicate? x)
  (bar x)
 (baz x))
```
Clojure style does not.
```
;; PREFERRED?
(if (predicate? x)
  (bar x)
  (baz x))
```
If there are any elif parts, then Hy's `if` is most similar to Clojure's `cond` or Arc's `if`.
```
;; PREFERRED?
(if (< n 0.0) "negative"
    (= n 0.0) "zero"
    (> n 0.0) "positive"
    "not a number")
```
Maybe this looks weird if there's only a then and else part:
```
;; OK?
(if (predicate? x) (bar x)
  (baz x))
```
Clojure's style guide [recommends](https://github.com/bbatsov/clojure-style-guide#short-forms-in-cond) that `cond` pairs be simple enough to fit on one line, or where not possible, separated with comments or blank lines. One could always factor out functions, I suppose, but do blocks may also be OK.